code/json.sls

```diff
@@ -30,172 +30,232 @@
+            (write-any v p)))
+        (when pretty?
+          (set! indent-level (- indent-level 1))
+          (display "\n" p)
+          (display-level p))
+        (display "}" p))
 
       (define (write-array a p)
-	(display "[" p)
-	(let ((need-comma #f))
-	  (for-each (lambda (v)
-		      (if need-comma
-			  (display ", " p)
-			  (set! need-comma #t))
-		      (write-any v p))
-		    a))
-	(display "]" p))
+        (display "[" p)
+        (when pretty?
+          (display "\n" p)
+          (set! indent-level (+ indent-level 1)))
+        (let ((need-comma #f))
+          (for-each (lambda (v)
+                      (if need-comma
+                          (if pretty?
+                              (display ",\n" p)
+                              (display "," p))
+                          (set! need-comma #t))
+                      (when pretty? (display-level p))
+                      (write-any v p))
+                    a))
+        (when pretty?
+          (set! indent-level (- indent-level 1))
+          (display "\n" p)
+          (display-level p))
+        (display "]" p))
 
       (define (write-any x p)
-	(cond
-	  ((or (string? x)
-	       (number? x)) (write x p))
-	  ((boolean? x) (display (if x "true" "false") p))
-	  ((symbol? x) (write (if (eq? x 'null) 'null (symbol->string x))
-			      p)) ;; for convenience
-	  ((null? x) (display "null" p))
-	  ((and (list? x)
-		(pair? (car x))
-		(not (pair? (caar x))))
-	   (write-ht x p))
-	  ((list? x) (write-array x p))
-	  (else (error "Invalid JSON object in json-write" x))))
+        (cond
+         ((or (string? x)
+              (number? x)) (write x p))
+         ((boolean? x) (display (if x "true" "false") p))
+         ((symbol? x)
+          (write (if (eq? x 'null) 'null (symbol->string x)) p))
+         ((null? x) (display "null" p))
+         ((and (list? x)
+               (pair? (car x))
+               (not (pair? (caar x))))
+          (write-ht x p))
+         ((list? x) (write-array x p))
+         (else (error "Invalid JSON object in json-write" x))))
 
-      (lambda (x . maybe-port)
-	(write-any x (if (pair? maybe-port) (car maybe-port) (current-output-port))))))
+      (case-lambda
+        ;; For default unformatted json rendering
+        ((obj)
+         (set! pretty? #f)
+         (write-any obj (current-output-port)))
+        ;; For default Pretty Printing with a tabstop of 2 spaces
+        ((msg obj)
+         (case msg
+           ((pretty)
+            (set! pretty? #t)
+            (set! tabstop-size 2)
+            (write-any obj (current-output-port)))
+           (else (error 'json-write "Invalid message" msg))))
+        ;; Pretty Print with custom tabsize defined in spaces
+        ((msg obj tabsize)
+         (case msg
+           ((pretty)
+            (set! pretty? #t)
+            (set! tabstop-size tabsize)
+            (write-any obj (current-output-port)))
+           (else (error 'json-write "invalid message" msg)))))))
 
```

Comment:
I think perhaps the arguments should be reversed? It's seems more natural if the object to print appears before the flag to pretty print. What do you think?

Reply:
Yeah, I am fine with that.

Reply:
Reversed them and fixed the documenting comment at the top